### PR TITLE
Test by TDD

### PR DESCRIPTION
**Pull request by HongJungKim-dev**
Add to test it is not empty

todo가 있는 경우를 테스트합니다.

### CODE REVIEW

**Comment by HongJungKim-dev** [comment]
33c4f3134b3f8d5f799ca31b136e7ada98f49788

**Comment by Kihyun92** [comment]
> Item.test.jsx에서 완료버튼에 대한 동작을 체크할 수 없으니 List.test.jsx에서는 그에 관한 테스트를 할 필요가 없다고 생각해서 포함시키지 않았습니다. 맞는 접근인지 궁금합니다.

굳이 상위에서 같은 테스트를 해야하지 않다면 실제 구현된 컴포넌트에서만 테스트를 해도 될 것 같아요. 근데 리스트에서 예를들어 두번째 리스트에 있는 할일의 완료 버튼을 눌렀을 경우에 어떻게 되어야하는지에 대한 테스트가 필요하다면 List.test.jsx에서도 테스트를 구현해주는게 좋겠네요.ㅎㅎ

> 그리고, onClickDelete를 포함하는지 검사하는것은 필요할것 같아서 그러한 검사를 하기 위한 함수를 찾아보다가 없어서 하지 못했습니다. 이 역시 onClickDelete함수를 포함해야 이후에 item.test.jsx에서 테스트가 가능해지므로 List.test.jsx에서는 테스트할 필요가 없는것인가요?

완료 버튼을 클릭함으로써 결국 `onClickDelete`를 호출하게 되는거 아닌가요? 맞다면 위에서 답변한 부분 같습니다 :)

혹시 잘못됐거나 더 의논이 필요하다면 말씀해주세요!

File: src/List.test.jsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import { render } from '@testing-library/react';
+
+import List from './List';
+
+test('EmptyList', () => {
```
Comment (Kihyun92) [inline src/List.test.jsx:6]: 오 좋습니다. 리스트가 비었을 경우와 비어있지 않는 경우에 대해서 각각 테스트 코드를 작성해보셨네요.

현재 이 테스트가 실제로 무엇을 하는지 더 자세히 설명해주면 더 좋을 것 같아요.

다만 이 테스트가 뭘 하려는지 조금 더 명확하게 표현해주면 좋을 것 같아요. 이러한 경우 `describe - context - it`의 구조로 표현해보는 건 어떤가요?

``` jest
describe('List', () => {
  context('test의 length가 0인 경우', () => {
    const tasks = [];

    it('"할 일이 없어요!"를 표시한다.', () => {
      const { container } = render((
        <List tasks={tasks} onClickDelete={handleClick} />
      ));

      expect(container).toHaveTextContent('할 일이 없어요!');
    });
  });

  context('test의 length가 1 이상인 경우', () => {
    ~~ 생략 ~~
```
일단 제 임의대로 작성해봤는데, 차례대로 보면 `List는 test의 length가 0인 경우 "할 일이 없어요!"를 표시한다.`와 같이 읽혀서 조금 더 명확해 진 것 같지 않나요? :)

참고
- https://ko.javascript.info/testing-mocha
-  https://www.betterspecs.org/#contexts

**Review by Kihyun92 — CHANGES_REQUESTED** [review]
빠르게 시작해주셨네요 👍🏼

File: src/List.test.jsx [hunk 2]
```diff
@@ -3,33 +3,39 @@ import { render } from '@testing-library/react';
...
+    const handleClick = jest.fn();
 
-  const handleClick = jest.fn();
+    it('"할 일이 없어요!"를 표시한다.', () => {
+      const { container } = render((
+        <List
+          tasks={tasks}
+          onClickDelete={handleClick}
+        />
+      ));
```
Comment (Kihyun92) [inline src/List.test.jsx:17]: 매 테스트마다 리스트 컴포넌트를 렌더하기 위해 중복 코드가 발생하는 것 같아요.

공통 함수를 `it` 바깥으로 분리할 수 있을까요?

**Review by Kihyun92 — CHANGES_REQUESTED** [review]

**Comment by HongJungKim-dev** [comment]
slack에 다른 분들도 비슷한 고민을 하면서 하는 것 같습니다. 저도 테스트를 하면서 이렇게 TDD 하는것이 맞는지, 어디서부터 어디까지 테스트 코드를 작성해야하는지, 이런것까지 테스트 코드를 작성해야하는지 고민이 들었는데 이미 완성한 코드에 대한 테스트 코드를 작성해서 그럴 수 있다는 것을 답변을 통해서 알았습니다.
그런데 스스로 왜 그러한 질문을 하지 않았을까를 생각했습니다. 
 첫번째로 조금이라도 더 커밋해서 코드리뷰를 받고 싶었던 것 같습니다. 그렇다고 커밋을 많이 하지도 못했습니다. 왜냐하면 커밋하는 과정도 Fail하고 Pass가 된 이후에야 Fail하는 코드로 되돌리고 Pass로 커밋하였습니다. 만약에 Pass가 안되면 Fail에서 어떻게 Pass해야할지 몰라서 찾아보고 혼자 고민하다가 커밋을 못했습니다. 전에는 미완성 코드라도 올리기라도 했는데 TDD는 Fail, Pass, Refactor 과정을 보실 수 있도록 차례대로 보여드려야한다는 생각이 강했던 것 같습니다. 이제 그냥 Fail나는 대로 바로 커밋을 해야겠다는 생각을 했습니다.
 두번째로 이런식으로 하는게 맞는지 고민하는 생각 자체가 TDD가 아직 익숙하지 않기때문이라않 생각했던 것 같습니다. 하지만 TDD 같은 생소한 개념은 더더욱 사소한것으로 생각하는 문제라도 질문해서 방향성을 제대로 잡는게 맞는 것 같습니다. 앞으로는 사소하게 고민이 드는게 있다면 그 생각을 정리해서 남겨야겠습니다.

**Comment by HongJungKim-dev** [comment]
되도록 run하기전에 어떤식으로 동작할것을 예상하고 짜도록 즉, 생각하고 코드를 작성해야한다고 이전에 생각했었습니다. 그런데 TDD는 Fail과 Pass를 반복하는 과정에서 생각하고 코드를 작성하는것보다 일단 빠르게 Fail 해서 오류 메시지를 보고 다시 Pass시키도록 하게 되는것이라면 TDD는 생각한대로 코드가 동작하는것보다 빠르게 작성하고 이를 수정하는것을 중요시 하는것인가요? 제가 TDD개념을 잘못 생각하는 것인가요?

**Comment by HongJungKim-dev** [comment]
List.test.jsx에서도 onClickDelete 테스트를 하는게 아직도 이해가 안갑니다.
어차피 Item.test.jsx에서 테스트를해서 통과여부가 결과로 나올텐데 List에서 중복으로 해야하는 이유가 있을까요? 제가 무언가를 놓치고있을까요? 다음 컴포넌트로 가기전에 그래도 이 컴포넌트에 있는 모든것은 검사해야해서 그런것일까요?

**Comment by Kihyun92** [comment]
> 이전에 코드를 작성하는 방법이 되도록 run하기전에 어떤식으로 동작할것을 예상하고 코드를 작성해야한다고 생각했었습니다. 그런데 TDD는 Fail과 Pass를 반복하는 과정에서 생각하고 코드를 작성하는것보다 일단 빠르게 Fail 해서 오류 메시지를 보고 다시 Pass시키도록 하게 되는것이라면 TDD는 생각한대로 코드가 동작하는것보다 빠르게 작성하고 이를 수정하는것을 중요시 하는것인가요? 제가 TDD개념을 잘못 생각하는 것인가요

TDD도 생각한 것을 빠르게 구현하기 위해 사용하는 것이죠. 다만 접근 방식이 다를 뿐, 뭘 더 중시한다 이런 개념이 아닌 것 같습니다. 코드숨 강의에 테스트 쪽 자료를 한번 보시면 좋을 것 같아요.(덧셈 함수 만들어보는 부분입니다)
덧셈을 해주는 기능을 만들려고 생각했을때, TDD를 사용하여 일단 실패하는 테스트 코드를 만들어보고,(1+3 = 4) 이걸 성공하도록 실제 코드를 구현하고, 리펙터링하고, 이 과정을 반복함으로써 우리가 생각한 기능을 구현하는 과정이에요.

TDD를 할때도 어떤식으로 동작할지 예상해야하는 부분은 같습니다. 바로 코드를 작성하느냐 아니면, 내가 뭘 확인해봐야하는지를 앞서 생각해보고, 연관된 테스트를 작성하며 구현하느냐 와 같이 `문제 해결을 위한 접근 방식이 다르다`고 볼 수 있겠네요.

일단 빠르게 fail하라. 이 말이 아무 테스트나 만들어보라는 말은 아니니깐요 ㅎㅎ
뭘 테스트하고 싶은건지. 목적을 한번 고민해보시면 좋을 것 같아요.

> Fail과 Pass과정에서 에러메시지를 보고 다음 코드를 작성하는것이 생각해서 코드를 작성하기보다 에러메시지 대로 수동적으로 작성하는것 아닌가요? TDD방식을 안좋게 얘기하려는게 아니라 잘 모르겠습니다.
그리고 test코드 작성으로 잡아내지 못하는 문제는 없을까요? 그건 개발자의 역량이고 그것을 키워야하는지 궁금합니다.

위에서 말씀드린대로 뭘 확인해야할지 모른 상태로 단순히 기능을 테스트한다면 에러 메시지대로 수동적으로 구현하게 되겠죠ㅎㅎ

test로 잡아내지 못하는 케이스가 있을 수 있습니다. 그렇기 때문에 계속해서 리팩터링을 하고 테스트를 추가해야합니다.

**Comment by Kihyun92** [comment]
> [33c4f31](https://github.com/CodeSoom/react-week3-assignment-1/commit/33c4f3134b3f8d5f799ca31b136e7ada98f49788)
> List.test.jsx에서도 onClickDelete 테스트를 하는게 아직도 이해가 안갑니다.
> 어차피 Item.test.jsx에서 테스트를해서 통과여부가 결과로 나올텐데 List에서 중복으로 해야하는 이유가 있을까요? 제가 무언가를 놓치고있을까요? 다음 컴포넌트로 가기전에 그래도 이 컴포넌트에 있는 모든것은 검사해야해서 그런것일까요?

그렇다면 테스트를 하지 않으셔도 될 것 같아요ㅎㅎ 
테스트에서 생각할 부분은 `내가 얼마나 안심할 수 있는가`라고 생각해요. 이 과정에서는 어느정도 중복이 발생할 수도 있습니다.

그리고 각 컴포넌트마다 가지고 있는 관심사의 수준이 다르기 때문에 하위 컴포넌트의 구현 상세를 과도하게 테스트하게 될 수 있습니다. 이 경우 상위에서 너무 많은 걸 알고 있다는 신호고 이는 설계적으로 문제가 될 수 있습니다.

그러나 이 부분에 대해서 지금 너무 깊게 고민하지 않으셨으면 해요. 경험치가 누적되지 않으면 사실 판단하기 어렵습니다. 그래서 훈련할 땐 극단적으로 테스트를 작성하는 걸 권합니다. 그래야 적정 수준의 테스트가 무엇인지 알 수 있게 됩니다. 지금은 떠오르는 불안들은 잠시 넣어두시고, 테스트 커버리지 100프로에 대한 집착만 남겨둬 보셔요🙂

**Review by Kihyun92 — CHANGES_REQUESTED** [review]

File: src/Input.test.jsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  expect(container).toHaveTextContent('추가');
+  expect(getByPlaceholderText('할 일을 입력해주세요')).toHaveTextContent('할 일을 입력해 주세요');
+
+  expect(handleClick).not.toBeCalled();
+  fireEvent.click(getByText('추가'));
+  expect(handleClick).toBeCalled();
+
+  it('입력이 변경됨을 표시합니다.', () => {
+    ReactTestUtils.Simulate.change(React.createRef);
+  });
```
Comment (Kihyun92) [inline src/Input.test.jsx:25]: 이 테스트의 의도는 무엇인가요?

File: src/List.test.jsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import React from 'react';
+import { render } from '@testing-library/react';
+
+import List from './List';
+
+describe('List', () => {
+  function renderList({ component }) {
+    return render(component);
+  }
+  context('test의 length가 0인 경우', () => {
```
Comment (Kihyun92) [inline src/List.test.jsx:10]: 코드 사이에 공백 추가해주세요~

그리고 `test의 length가 0인 경우`라는게 tasks가 없는 경우를 의도하신게 맞나요? ㅎㅎ

맞다면 단순하게 `task가 없는 경우`와 같이 표현해주는건 어떤가요? 🤔

**Review by Kihyun92 — CHANGES_REQUESTED** [review]
많이 고민해보며 진행중이신 것 같아요! 💯 

의논이 필요하다면 언제든 말씀주세요~ 👍

**Comment by HongJungKim-dev** [comment]
> TDD도 생각한 것을 빠르게 구현하기 위해 사용하는 것이죠. 다만 접근 방식이 다를 뿐, 뭘 더 중시한다 이런 개념이 아닌 것 같습니다. 코드숨 강의에 테스트 쪽 자료를 한번 보시면 좋을 것 같아요.(덧셈 함수 만들어보는 부분입니다)
덧셈을 해주는 기능을 만들려고 생각했을때, TDD를 사용하여 일단 실패하는 테스트 코드를 만들어보고,(1+3 = 4) 이걸 성공하도록 실제 코드를 구현하고, 리펙터링하고, 이 과정을 반복함으로써 우리가 생각한 기능을 구현하는 과정이에요.

**Comment by HongJungKim-dev** [comment]
## Affirmation

TDD 과제가 어려워서 정말 집중이 안되어서 커밋한 양이 별로 되지 않습니다.  뭘 테스트해야할지 모르고 익숙하지 않다면 그냥 테스트할 수 있는것은 무엇이든 테스트해보려고 합니다. 비록 9시가 지나가겠지만 계속해서 포기하지 않고 더 커밋하고 싶습니다.

**Comment by HongJungKim-dev** [comment]
> 그러나 이 부분에 대해서 지금 너무 깊게 고민하지 않으셨으면 해요. 경험치가 누적되지 않으면 사실 판단하기 어렵습니다. 그래서 훈련할 땐 극단적으로 테스트를 작성하는 걸 권합니다. 그래야 적정 수준의 테스트가 무엇인지 알 수 있게 됩니다. 지금은 떠오르는 불안들은 잠시 넣어두시고, 테스트 커버리지 100프로에 대한 집착만 남겨둬 보셔요🙂

말씀하신대로 과도하게 테스트한다는 생각으로 List.test.jsx도 onClickDelete를 test하는것을 추가했습니다.

**Comment by HongJungKim-dev** [comment]
## 출력 테스트 질문

Input.test.jsx를 하면서 입력창과 추가버튼을 출력하는것도 출력해야하는지 의문이 들었습니다. 그래서 Item.test.jsx를 다시 살펴봤는데 출력하는것을 테스트하지는 않는것 같습니다. 이 부분은 테스트하지 않는것인지 궁금합니다.

## 다중 테스트 질문
Item.test.jsx에서는 하나의 데이터로 테스트를 진행했는데 1개 이상의 복수 데이터로 테스트 하는것을 추가하고 다른 테스트 파일에서도 동일하게 진행해야하는지 궁금합니다. 더 많은 데이터로 하는게 더 정확하고 이를 위해서는 코드가 일부 수정되어야한다고 생각했습니다.

**Comment by Kihyun92** [comment]
> ## 다중 테스트 질문
> Item.test.jsx에서는 하나의 데이터로 테스트를 진행했는데 1개 이상의 복수 데이터로 테스트 하는것을 추가하고 다른 테스트 파일에서도 동일하게 진행해야하는지 궁금합니다. 더 많은 데이터로 하는게 더 정확하고 이를 위해서는 코드가 일부 수정되어야한다고 생각했습니다.

많은 고민들을 해보시면서 하는 것 같아 좋습니다 ㅎㅎ 

그런데 Item 컴포넌트의 테스트에서 여러 데이터로 테스트해야한다고 생각하신 이유가 무엇인가요?

Item 컴포넌트는 할일의 id, title 정보를 받아와서 화면에 뿌려주는 역할이죠. 그런데 복수의 데이터로 테스트가 왜 필요하다고 생각하신 건지 궁금합니다.

복수의 task가 있을 경우에 대한 확인은 List 컴포넌트 테스트에서 진행하는게 맞지 않을까요?

File: src/Input.test.jsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  expect(container).toHaveTextContent('할 일');
+  expect(container).toHaveTextContent('추가');
+  expect(getByPlaceholderText('할 일을 입력해 주세요')).toBe('할 일을 입력해 주세요');
+
+  expect(handleClick).not.toBeCalled();
+  fireEvent.click(getByText('추가'));
+  expect(handleClick).toBeCalled();
+
+  it('입력이 변경됨을 표시합니다.', () => {
+    fireEvent.change(getByPlaceholderText('할 일을 입력해 주세요'), { target: { value: 'TDD 과제하기' } });
```
Comment (Kihyun92) [inline src/Input.test.jsx:23]: onChange 이벤트에 대한 테스트를 해보신 것 같아요~

그런데 이벤트를 발생시킨 부분만 있는 것 같아요. 어떤걸 expect 해야할까요??

**Review by Kihyun92 — CHANGES_REQUESTED** [review]
아직 작업을 더 진행해보고 계신 것 같아요ㅎㅎ

계속 고민해보고 시도해보시는게 너무 좋습니다 👍🏼

File: src/List.test.jsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        handleClickDelete={handleClickDelete}
+      />
+    ));
+  }
+
+  context('task가 없을 경우', () => {
+    const tasks = [];
+
+    it('"할 일이 없어요!"를 표시한다.', () => {
+      const { container } = renderList({ tasks, handleClick });
```
Comment (Kihyun92) [inline src/List.test.jsx:22]:
```suggestion
      const { container } = renderList({ tasks, handleClickDelete: handleClick });
```
맵핑되는 함수명이 달라서 에러가 발생하는 것 같아요. 지금과 같이 함수명을 다른 상황에서 객체로 넘기는 경우엔 이렇게 맵핑 해줘야할 것 같네요.

**Review by Kihyun92 — CHANGES_REQUESTED** [review]

File: src/List.test.jsx (hunk 6 above)
Comment (Kihyun92) [reply src/List.test.jsx:22]: 구조분해할당에 대해서 봐보시면 도움 될 것 같아요~

- https://ko.javascript.info/destructuring-assignment#ref-1545

**Comment by HongJungKim-dev** [comment]
> 많은 고민들을 해보시면서 하는 것 같아 좋습니다 ㅎㅎ
그런데 Item 컴포넌트의 테스트에서 여러 데이터로 테스트해야한다고 생각하신 이유가 무엇인가요?
Item 컴포넌트는 할일의 id, title 정보를 받아와서 화면에 뿌려주는 역할이죠. 그런데 복수의 데이터로 테스트가 왜 필요하다고 생각하신 건지 궁금합니다.
복수의 task가 있을 경우에 대한 확인은 List 컴포넌트 테스트에서 진행하는게 맞지 않을까요?

말씀하신 것처럼 Item은 뿌려주는 역할이고 여러 테스트는 List에서 하는게 맞는것 같습니다. 그렇다면 상위 컴포넌트인 Page에서 task가 있을경우에서 여러 테스트를 진행하는게 맞을것 같습니다.

File: src/Input.test.jsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+  expect(container).toHaveTextContent('할 일');
+  expect(container).toHaveTextContent('추가');
+  expect(getByPlaceholderText('할 일을 입력해 주세요')).toBe('할 일을 입력해 주세요');
+
+  expect(handleClick).not.toBeCalled();
+  fireEvent.click(getByText('추가'));
+  expect(handleClick).toBeCalled();
+
+  context('사용자가 Todo를 입력하는 경우', () => {
```
Comment (Kihyun92) [inline src/Input.test.jsx:25]: `context` 보다는 `describe`로 처리해주는게 좋을 것 같네요~

`with <-> without`
`when log in <-> when log out`
처럼 같은 맥락에서 별개의 사례일 때 `context`를 사용합니다.

context 사용법은 아래 문서와 아샬님의 문서를 참고해주세요.

- https://www.betterspecs.org/
- https://github.com/ahastudio/til/blob/main/blog/2018/12-08-given-when-then.md

File: src/Input.test.jsx [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    it('입력이 변경됨을 표시합니다.', () => {
```
Comment (Kihyun92) [inline src/Input.test.jsx:26]: `it`으로 별개의 테스트를 추가하려는 의도 같아요.

`it` 위쪽 테스트 부분이 별개의 테스트라면 따로 묶어 처리하는 건 어떤가요?

File: src/List.test.jsx [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  const handleClick = jest.fn();
+
+  function renderList({ tasks, handleClickDelete }) {
+    return render((
+      <List
+        tasks={tasks}
+        onClickDelete={handleClickDelete}
+      />
+    ));
+  }
```
Comment (Kihyun92) [inline src/List.test.jsx:16]:
```suggestion
  const handleClickDelete = jest.fn();

  function renderList(tasks) {
    return render((
      <List
        tasks={tasks}
        onClickDelete={handleClickDelete}
      />
    ));
  }
```

굳이 호출시점에 핸들러 함수를 넘길 필요 없이 이렇게 처리하면 인자 하나를 없앨 수 있을 것 같아요 ☺️

File: src/List.test.jsx [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        id: 1,
+        title: '뭐라도 하기',
+      },
+    ];
+
+    it('입력된 할 일을 표시한다.', () => {
+      const { container } = renderList({ tasks, handleClickDelete: handleClick });
+
+      expect(container).toHaveTextContent('뭐라도 하기');
+    });
```
Comment (Kihyun92) [inline src/List.test.jsx:40]: 만약 tasks가 여러개 있는 경우에는 어떻게 테스트 할 수 있을까요?

File: src/Page.test.jsx [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+      const { container } = renderPage({ taskTitle, tasks });
+
+      expect(container).toHaveTextContent('할 일이 없어요!');
+      expect(getByPlaceholderText('할 일을 입력해 주세요')).toBe('할 일을 입력해 주세요');
+      expect(getByPlaceholderText('할 일을 입력해 주세요')).toValue('');
+    });
+
+    expect(onClickAddTask).not.toBeCalled();
+    fireEvent.click(getByText('추가'));
+    expect(onClickAddTask).toBeCalled();
```
Comment (Kihyun92) [inline src/Page.test.jsx:42]: 테스트 하는 부분은 it 함수 안에서 처리해주는게 좋을 것 같네요~

File: src/Page.test.jsx [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  function renderPage({ taskTitle, tasks }) {
+    return render((<Page
+      taskTitle={taskTitle}
+      tasks={tasks}
+      onClickAddTask={onClickAddTask}
+      onClickDeleteTask={onClickDeleteTask}
+    />));
+  }
+
+  const { mainContainer } = render((
```
Comment (Kihyun92) [inline src/Page.test.jsx:19]: `mainContainer`는 무엇인가요? `container`를 사용하지 않는 이유가 있나요?

```suggestion
  const { container } = render((
```

File: src/Page.test.jsx [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    const tasks = [
+      { id: 1, title: '공부하기' },
+    ];
+
+    it('입력된 값과 할 일들을 보여준다.', () => {
+      const { container, getByText, getByPlaceholderText } = renderPage({ taskTitle, tasks });
+
+      expect(container).toHaveTextContent('할 일이 없어요!');
+      expect(getByPlaceholderText('할 일을 입력해 주세요')).toBe('할 일을 입력해 주세요');
+      expect(getByPlaceholderText('할 일을 입력해 주세요')).toValue(taskTitle);
```
Comment (Kihyun92) [inline src/Page.test.jsx:56]: toValue는 어떤 의도일까요?

**Review by Kihyun92 — CHANGES_REQUESTED** [review]

File: src/Page.test.jsx [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    return render((<Page
+      taskTitle={taskTitle}
+      tasks={tasks}
+      onClickAddTask={onClickAddTask}
+      onClickDeleteTask={onClickDeleteTask}
+    />));
+  }
+
+  it('To-do와 할 일을 표시한다.', () => {
+    const { container } = renderPage();
```
Comment (Kihyun92) [inline src/Page.test.jsx:20]: renderPage 함수 호출시 인자로 초기 값을 넘기거나 혹은 함수에서 default value에 대한 처리가 있어야 할 것 같아요~

File: src/Page.test.jsx [hunk 15]
```diff
@@ -0,0 +1,79 @@
+import React from 'react';
+import { render, fireEvent } from '@testing-library/react';
+
+import Page from './Page';
+
+describe('Page', () => {
+  const onClickAddTask = jest.fn();
+  const onClickDeleteTask = jest.fn();
```
Comment (Kihyun92) [inline src/Page.test.jsx:8]: 현재 mocking한 onClickAddTask 함수를 각 테스트마다 호출하고 있는 것 같아요.

지금 구조라면 테스트마다 side effect가 생길 수 있습니다. 의도치 않은 영향이 생겨 테스트가 우리 의도대로 동작하지 않을 수 있는거죠.

각 테스트는 독립적으로 실행되어야합니다. 
이 경우에는 beforeEach`를 이용해서 mocking 함수들을 초기화 시켜주면 각 테스트가 의도치 않게 동작하는 것을 막을 수 있습니다.

```jest
beforeEach(() => {
  jest.clearAllMocks();
});
```
- [https://jestjs.io/docs/en/jest-object#jestclearallmocks](https://jestjs.io/docs/en/jest-object#jestclearallmocks)
- [https://jestjs.io/docs/en/setup-teardown](https://jestjs.io/docs/en/setup-teardown)

**Review by Kihyun92 — CHANGES_REQUESTED** [review]
많은 시도를 해보시는게 좋네요 👍🏼

**Comment by HongJungKim-dev** [comment]
## 질문

App.test.jsx에서 추가버튼과 완료버튼에 해당하는 onChange나 onClick함수가 호출됬는지 체크할 수 없어서 input이 changed되는 것을 container가 갖고 있는지나 완료버튼에 해당하는것들은 관심사가 아니라고 판단해서 테스트 하지 않았습니다.  완료버튼도 추가가되어서 내용이 변경되어야 테스트가 가능하기 때문입니다. 맞는 방향인지 궁금합니다.

File: src/Input.test.jsx [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+import Input from './Input';
+
+describe('Input', () => {
+  const handleClick = jest.fn();
+  const onChange = jest.fn();
+
+  function renderInput({ value }) {
+    return render((<Input
+      value={value}
+      handleClick={handleClick}
```
Comment (Kihyun92) [inline src/Input.test.jsx:14]: Input 컴포넌트에서는 onClick으로 받고 있는 것 같아요.

```suggestion
      onClick={handleClick}
```

File: src/Input.test.jsx [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    expect(handleClick).toBeCalled();
+  });
+
+  it('입력이 변경됨을 표시합니다.', () => {
+    const { getByPlaceholderText } = renderInput({ value: 'TDD 과제하기' });
+
+    expect(getByPlaceholderText('할 일을 입력해 주세요')).toHaveValue('TDD 과제하기');
+
+    expect(onChange).not.toBeCalled();
+    fireEvent.change(getByPlaceholderText('할 일을 입력해 주세요'), { target: { value: 'TDD 과제하기' } });
```
Comment (Kihyun92) [inline src/Input.test.jsx:42]: 기존의 값과 변경할 값이 동일하기 때문에 onChange가 호출이 안될 것 같네요

File: src/Input.test.jsx [hunk 18]
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+
+import { fireEvent, render } from '@testing-library/react';
+
+import Input from './Input';
+
+describe('Input', () => {
+  const handleClick = jest.fn();
+  const onChange = jest.fn();
```
Comment (Kihyun92) [inline src/Input.test.jsx:9]: 네이밍을 onChange가 아닌 handleChange는 어떤가요?

**Review by Kihyun92 — CHANGES_REQUESTED** [review]

**Comment by HongJungKim-dev** [comment]
commit db2960d6602d300d898643664e3081587e5d915d

**Comment by Kihyun92** [comment]
> ## 질문
> [db2960d](https://github.com/CodeSoom/react-week3-assignment-1/commit/db2960d6602d300d898643664e3081587e5d915d)
> 
> fireEvent.change로 변경된것을 확인하고 싶은데 왜 초기값 그대로 ''이어서 기대한 'TDD 과제하기'가 안나오는지 궁금합니다.

변경 되지 않은 값으로 change를 요청한 경우, onChange가 호출되지 않기 때문입니다. ☺️

**Review by Kihyun92 — APPROVED** [review]
고생하셨습니다 ㅎㅎ 

이번 과제를 해보시면서 많은 고민을 해보신 것 같아요. 질문도 많이 해주셔서 덕분에 저도 많이 고민할 수 있었습니다 👍🏼 

혹시 더 의논이 필요한 부분이 있다면 말씀해주시고, 과제 풀이 영상도 꼭 보시면 좋겠습니다.

이번 과제를 통해 무엇을 배우셨나요? 더 배우고 싶은 부분이 있나요? 😊

슬랙 채널에 회고 남겨주세요~

**Comment by HongJungKim-dev** [comment]
100% 커버리지 보다 테스트를 더 추가하고 있는데 얼마나 과도하게 테스트하느냐는 테스트하는 사람의 입장에 따라 다른 것인가요 아니면 100%를 달성해도 중요한 테스트를 놓칠 수 있는것인가요? 후자일 가능성도 있으니 항상 과도하게 테스트 해야하는것인지 궁금합니다.

**Comment by hannut91** [comment]
얼마나 테스트할 것인지는 결정해야 합니다. 얼마나 믿고 기능을 수정할 수 있는지는 사람마다 다를 것 같아요.
테스트 커버리지가 100%를 달성하더라도 놓친 테스트가 있을 수 있습니다. 왜냐하면 테스트 커버리지가 테스트의 품질을 보장하지 않기 때문입니다.

다음 글을 꼭 읽어보세요.

See also
  - https://martinfowler.com/bliki/TestCoverage.html

**Comment by HongJungKim-dev** [comment]
>>얼마나 테스트할 것인지는 결정해야 합니다. 얼마나 믿고 기능을 수정할 수 있는지는 사람마다 다를 것 같아요.
테스트 커버리지가 100%를 달성하더라도 놓친 테스트가 있을 수 있습니다. 왜냐하면 테스트 커버리지가 테스트의 품질을 보장하지 않기 때문입니다.

>>If it seems like a simple change to code causes excessively long changes to tests, that's a sign that there's a problem with the tests. This may not be so much that you are testing too many things, but that you have duplication in your tests.

참고링크
* https://martinfowler.com/bliki/TestCoverage.html

해당 링크를 읽고 결국 얼마나 테스트하는것이 저에게 달려 있고, 코드를 추가할때 테스트 코드가 지나치게 길어지면 테스트에 문제가 있다는 것을 알았습니다.
그런데 이해가 안가는것이 있습니다. 코드를 추가할때 테스트 코드가 지나치게 길어진다는것은 3주차 처럼 코드를 작성하고 이후에 테스트 코드를 작성하는 것과 같은 상황으로 이해했습니다. 
하지만 이 글은 TDD와 같은 테스트 할때를 얘기하기 때문에 테스트코드 작성 -> 코드 작성 -> 리팩토링 -> 테스트코드 작성... 와 같은 반복된 과정에서 코드 작성 이후 다른 부분 테스트 코드 작성시에 지나치게 길어지면 문제가 있다는것을 뜻하는 것인지 모르겠습니다.
이번주부터 TDD cylce로 과제를 수행하면서 느껴봐야 알 수 있을까요?

**Comment by HongJungKim-dev** [comment]
aa9c6e4b99ea04aea4bcfe1d934ff38b6ce718e8

render에 값을 초기화하는 경우에는 Fail하는데 이 경우를 테스트하려면 어떻게 할지 모르겠습니다.

**Comment by Kihyun92** [comment]
> 해당 링크를 읽고 결국 얼마나 테스트하는것이 저에게 달려 있고, 코드를 추가할때 테스트 코드가 지나치게 길어지면 테스트에 문제가 있다는 것을 알았습니다.
그런데 이해가 안가는것이 있습니다. 코드를 추가할때 테스트 코드가 지나치게 길어진다는것은 3주차 처럼 코드를 작성하고 이후에 테스트 코드를 작성하는 것과 같은 상황으로 이해했습니다.
하지만 이 글은 TDD와 같은 테스트 할때를 얘기하기 때문에 다시 생각해봤습니다. 테스트코드 작성 -> 코드 작성 -> 리팩토링 -> 테스트코드 작성... 와 같은 반복된 과정에서 코드 작성 이후 다른 부분 테스트 코드 작성시에 지나치게 길어지면 문제가 있다는것을 뜻하는 것인가요?
이번주부터 TDD cylce로 과제를 수행하면서 느껴봐야 알 수 있을까요?

음 제 생각에는 테스트 코드가 지나치게 길어지면 테스트에 문제가 있다는 것은 절대적인 테스트 코드의 양이 많다고 해서 문제라는 것 보단 불필요한 테스트를 포함하고 있을 수 있기 때문에 문제라는 의미일 것 같아요. 

따라서 3주차에서 이미 작성된 테스트에 테스트 코드를 작성하는 경우라고 해서 지나치게 테스트가 길어진다고 볼 수 없죠. 마찬가지로 TDD cycle로 코드를 작성한다고 해서 지나치게 테스트가 길어지지 않을 보장이 되는 것도 아닐 것 같구요. 

물론 TDD cycle로 개발하면 핵심 테스트만 먼저 작성할 수 있기 때문에 불필요한 테스트들을 작성하지 않을 수 있을 겁니다. 하지만 윤석님 말처럼 얼마나 테스트 할 것인지가 결정되지 않았다면, 마찬가지로 지나친 테스트 코드가 작성 될 수 있을 것 같아요.

안심할 수 있는 테스트들을 작성했고, 얼마나 테스트할 것인지가 결정이 됐다면 지나치게 길어지는 테스트 코드가 작성되기 힘들 것 같아요~

지금은 혼란스러울 수 있지만, 앞으로 TDD cycle을 통해 과제들을 진행하면서 고민해보셔도 좋을 것 같습니다 ☺️

**Comment by Kihyun92** [comment]
> [aa9c6e4](https://github.com/CodeSoom/react-week3-assignment-1/commit/aa9c6e4b99ea04aea4bcfe1d934ff38b6ce718e8)
> 
> render에 값을 초기화하는 경우에는 Fail하는데 이 경우를 테스트하려면 어떻게 할지 모르겠습니다.

어쩌면 테스트하는 방식이 잘못된 걸 수도 있을 것 같아요.

해당 테스트에서 render에 값을 초기화하는 경우에는 Fail하는 원인을 먼저 알아봐야 할 것 같습니다.🤔

상위 컴포넌트에서 input을 테스트할때, 명시적으로 render 함수에 value를 넘기는 경우에 value 값이 제대로 업데이트가 되지 않는 건지는 저도 찾아보는 중입니다.

**Comment by HongJungKim-dev** [comment]
말씀하신대로 충분한 테스트 코드 작성하는것에 대한것은 4주차 과제하면서 고민해보겠습니다.
